### PR TITLE
Add SVM oneAPI interface tests

### PR DESCRIPTION
**Pull request by vlad-nazarov**

### CODE REVIEW

**Comment by vlad-nazarov** [comment]
http://intel-ci.intel.com/eb7bbfa3-4fd5-f112-9658-8cdcd4b723a1

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify linear not separable surface",
+                     "[svm][integration][batch]",
+                     svm_types) {
+    SKIP_IF(this->not_available_on_device());
+
+    using Float = std::tuple_element_t<0, TestType>;
+    using Method = std::tuple_element_t<1, TestType>;
+    using KernelTypeLinear = typename linear::descriptor<Float, linear::method::dense>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:372]:
```suggestion
    using float_t = std::tuple_element_t<0, TestType>;
    using method_t = std::tuple_element_t<1, TestType>;
    using kernel_t = linear::descriptor<Float, linear::method::dense>;
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+                                                   { 0.0, 1.0 } }),
+                                      LabelsPair({ {
+                                                       0.0,
+                                                       0.0,
+                                                       0.0,
+                                                       2.0,
+                                                       2.0,
+                                                       2.0,
+                                                   },
+                                                   { 0.0, 2.0 } }),
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:513]: Need to figure out how to make it more readable.

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+
+    using Float = std::tuple_element_t<0, TestType>;
+    using Method = std::tuple_element_t<1, TestType>;
+    using KernelTypeLinear = typename linear::descriptor<Float, linear::method::dense>;
+
+    constexpr std::int64_t row_count_train = 6;
+    constexpr std::int64_t column_count = 2;
+    constexpr std::int64_t element_count_train = row_count_train * column_count;
+
+    using WeightsPair = std::pair<std::array<Float, row_count_train>, std::array<Float, 1>>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:561]:
```suggestion
    using weights_pair_t = std::pair<std::array<Float, row_count_train>, std::array<Float, 1>>;
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify any two labels",
+                     "[svm][integration][batch]",
+                     svm_types) {
+    SKIP_IF(this->not_available_on_device());
+
+    using Float = std::tuple_element_t<0, TestType>;
+    using Method = std::tuple_element_t<1, TestType>;
+    using KernelTypeLinear = typename linear::descriptor<Float, linear::method::dense>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:475]:
```suggestion
    using float_t = std::tuple_element_t<0, TestType>;
    using method_t = std::tuple_element_t<1, TestType>;
    using kernel_t = linear::descriptor<Float, linear::method::dense>;
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+                           svm_desc,
+                           support_vector_count,
+                           support_indices,
+                           homogen_table{},
+                           y);
+}
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify any two labels",
+                     "[svm][integration][batch]",
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:469]:
```suggestion
                     "[svm][integration][batch][linear]",
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+
+    using Float = std::tuple_element_t<0, TestType>;
+    using Method = std::tuple_element_t<1, TestType>;
+    using KernelTypeLinear = typename linear::descriptor<Float, linear::method::dense>;
+
+    constexpr std::int64_t row_count_train = 6;
+    constexpr std::int64_t column_count = 2;
+    constexpr std::int64_t element_count_train = row_count_train * column_count;
+
+    using LabelsPair = std::pair<std::array<Float, row_count_train>, std::array<Float, 2>>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:481]:
```suggestion
    using labels_pair_t = std::pair<std::array<Float, row_count_train>, std::array<Float, 2>>;
```

Comment (rlnx) [reply cpp/oneapi/dal/algo/svm/test/batch.cpp:481]: Also suggest to move type aliases to the beginning of method where others are defined

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+                              svm_desc,
+                              support_vector_count,
+                              support_indices,
+                              decision_function,
+                              labels);
+}
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify linear separable surface with big margin",
+                     "[svm][integration][batch]",
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:309]:
```suggestion
                     "[svm][integration][batch][linear]",
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+                              svm_desc,
+                              support_vector_count,
+                              support_indices,
+                              decision_function,
+                              labels);
+}
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify linear not separable surface",
+                     "[svm][integration][batch]",
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:366]:
```suggestion
                     "[svm][integration][batch][linear]",
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+                              svm_desc,
+                              support_vector_count,
+                              support_indices,
+                              decision_function,
+                              labels);
+}
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify quadric separable surface with rbf kernel",
+                     "[svm][integration][batch]",
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:422]:
```suggestion
                     "[svm][integration][batch][rbf]",
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+                                 y,
+                                 svm_desc,
+                                 support_vector_count,
+                                 support_indices,
+                                 labels.second.data());
+}
+
+TEMPLATE_LIST_TEST_M(svm_batch_test,
+                     "svm can classify linear separable surface with weights",
+                     "[svm][integration][batch]",
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/svm/test/batch.cpp:549]:
```suggestion
                     "[svm][integration][batch][linear]",
```

File: cpp/oneapi/dal/algo/svm/test/batch.cpp (hunk 6 above)
Comment (vlad-nazarov) [reply cpp/oneapi/dal/algo/svm/test/batch.cpp:481]: In this case I use row_count_train. Should I remove all constexpr (row_count / column_count) and use only numerical values?

**Review by KulikovNikita — CHANGES_REQUESTED** [review]

File: cpp/oneapi/dal/algo/svm/common.cpp [hunk 11]
```diff
@@ -106,7 +106,7 @@ void descriptor_base<Task>::set_max_iteration_count_impl(std::int64_t value) {
 
 template <typename Task>
 void descriptor_base<Task>::set_cache_size_impl(double value) {
-    if (value <= 0.0) {
+    if (value < 0.0) {
         throw domain_error(dal::detail::error_messages::cache_size_leq_zero());
     }
```
[the remark this answers is not in the document]
Comment (vlad-nazarov) [reply cpp/oneapi/dal/algo/svm/common.cpp:111]: We discussed with @PetrovKP. Need an additional approve?

[the remark this answers is not in the document]
Comment (PetrovKP) [reply cpp/oneapi/dal/algo/svm/common.cpp:111]: @KulikovNikita  These are not interface changes. There are no additional approvals for this.

File: cpp/oneapi/dal/algo/svm/common.hpp [hunk 12]
```diff
@@ -118,7 +118,7 @@ class descriptor_base : public base {
     double get_cache_size() const;
 
     /// The parameter of the WSS scheme $\\tau$.
-    /// @invariant :expr:`tau >= 0.0`
+    /// @invariant :expr:`tau > 0.0`
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/svm/common.hpp:121]: Need to update message https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/detail/error_messages.cpp#L156

File: cpp/oneapi/dal/algo/svm/common.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (PetrovKP) [reply cpp/oneapi/dal/algo/svm/common.cpp:111]: Need to update error message https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/detail/error_messages.cpp#L144

**Review by PetrovKP — COMMENTED** [review]
Good job!

**Review by PetrovKP — APPROVED** [review]

**Comment by vlad-nazarov** [comment]
http://intel-ci.intel.com/eb8095bf-73ec-f1fb-a122-8cdcd4b723a1

**Comment by vlad-nazarov** [comment]
http://intel-ci.intel.com/eb809ff5-4e59-f1dd-bb84-8cdcd4b723a1

**Review by KulikovNikita — APPROVED** [review]

**Pull request merged**